x/config/chooser.go

```diff
@@ -0,0 +1,172 @@
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"go.uber.org/yarpc/api/peer"
+	peerbind "go.uber.org/yarpc/peer"
+)
+
+// PeerListConfig facilitates decoding and building peer choosers.
+// The peer chooser combines a peer list (for the peer selection strategy, like
+// least-pending or round-robin) with a peer list binder (like static peers or
+// dynamic peers from DNS or watching a file in a particular format).
+type PeerListConfig struct {
+	peerListConfig
+}
+
+// peerListConfig is the private representation of PeerListConfig that captures
+// decoded configuration without revealing it on the public type.
+type peerListConfig struct {
+	Peer string       `config:"peer,interpolate"`
+	Etc  attributeMap `config:",squash"`
+}
+
+// BuildChooser translates a chooser configuration into a peer chooser, backed
+// by a peer list bound to a peer list binder.
+func (pc PeerListConfig) BuildChooser(transport peer.Transport, identify func(string) peer.Identifier, kit *Kit) (peer.Chooser, error) {
+	c := pc.peerListConfig
+	// Establish a peer selection strategy.
+
+	// Special case for single-peer outbounds.
+	if c.Peer != "" {
+
+		if len(c.Etc) > 0 {
+			return nil, fmt.Errorf("unrecognized attributes in peer list config: %+v", c.Etc)
+		}
+
+		return peerbind.NewSingle(identify(c.Peer), transport), nil
+	}
+
+	// All multi-peer choosers may combine a peer list (for sharding or
+	// load-balancing) and a peer list updater.
+
+	// Find a property name that corresponds to a peer chooser/list and construct it.
+	for peerListName := range c.Etc {
+		peerListSpec := kit.peerListSpec(peerListName)
+		if peerListSpec == nil {
+			continue
+		}
+
+		var peerListUpdaterConfig attributeMap
+		if _, err := c.Etc.Pop(peerListName, &peerListUpdaterConfig); err != nil {
+			return nil, err
+		}
+
+		peerListUpdater, err := buildPeerListUpdater(peerListUpdaterConfig, identify, kit)
+		if err != nil {
+			return nil, err
+		}
+
+		chooserBuilder, err := peerListSpec.PeerList.Decode(c.Etc)
+		if err != nil {
+			return nil, err
+		}
+
+		result, err := chooserBuilder.Build(transport, kit)
+		if err != nil {
+			return nil, err
+		}
+
+		peerList := result.(peer.ChooserList)
+
+		return peerbind.Bind(peerList, peerListUpdater), nil
+	}
+
+	return nil, fmt.Errorf(
+		"no recognized peer list in config: got %s; need one of %s",
+		strings.Join(c.names(), ", "),
+		strings.Join(kit.peerListSpecNames(), ", "),
+	)
+}
+
+func (c peerListConfig) names() (names []string) {
+	for name := range c.Etc {
+		names = append(names, name)
+	}
+	sort.Strings(names)
+	return
+}
+
+func buildPeerListUpdater(c attributeMap, identify func(string) peer.Identifier, kit *Kit) (peer.Binder, error) {
+	var peers []string
+	_, err := c.Pop("peers", &peers)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(peers) > 0 {
+
+		if len(c) > 0 {
+			return nil, fmt.Errorf("unrecognized attributes in peer list config: %+v", c)
+		}
+
+		return peerbind.BindPeers(identifyAll(identify, peers)), nil
```

Comment:
We can add a "peers" spec that is implicitly always registered in a subsequent change for equivalent behavior without breaking backward compatibility.